File: web/server/canvas/operations/shared.ts
import type { CanvasNode, CanvasSnapshot } from "../types";

export type CanvasToolRequest = { name: "canvas_apply_ops"; input: Record<string, unknown> };
export type CanvasToolHandler = (input: Record<string, unknown>, state: CanvasSnapshot | null) => CanvasToolRequest;

/** 创建统一的批量画布操作请求。 */
export function applyOps(ops: unknown[]): CanvasToolRequest {
    return { name: "canvas_apply_ops", input: { ops } };
}

/** 创建文本节点操作。 */
export function textNodeOp(input: { id?: string; text?: string; title?: string; width?: number; height?: number }, x: number, y: number) {
    return { type: "add_node", id: input.id, nodeType: "text", title: input.title, position: { x, y }, width: input.width, height: input.height, metadata: { content: input.text || "", status: "success", fontSize: 14 } };
}

/** 创建生成配置节点操作。 */
export function configNodeOp(id: string, input: Record<string, unknown>, x: number, y: number) {
    const mode = generationMode(input.mode);
    const prompt = String(input.prompt || "");
    return {
        type: "add_node",
        id,
        nodeType: "config",
        title: String(input.title || generationTitle(mode)),
        position: { x, y },
        width: typeof input.width === "number" ? input.width : undefined,
        height: typeof input.height === "number" ? input.height : undefined,
        metadata: cleanRecord({
            generationMode: mode,
            composerContent: prompt,
            prompt,
            status: "idle",
            model: input.model,
            size: input.size,
            quality: input.quality,
            count: input.count,
            seconds: input.seconds,
            vquality: input.vquality,
            generateAudio: input.generateAudio,
            watermark: input.watermark,
            videoMode: input.videoMode,
            audioVoice: input.audioVoice,
            audioFormat: input.audioFormat,
            audioSpeed: input.audioSpeed,
            audioInstructions: input.audioInstructions,
        }),
    };
}

/** 创建触发节点生成的画布操作。 */
export function runGenerationOp(nodeId: string, mode: "text" | "image" | "video" | "audio", prompt?: string) {
    return { type: "run_generation", nodeId, mode, prompt };
}

/** 将未知生成模式归一为画布支持的模式。 */
export function generationMode(value: unknown): "text" | "image" | "video" | "audio" {
    return value === "text" || value === "video" || value === "audio" ? value : "image";
}

/** 获取生成模式对应的默认节点标题。 */
export function generationTitle(mode: "text" | "image" | "video" | "audio") {
    if (mode === "text") return "文本生成";
    if (mode === "video") return "视频生成";
    if (mode === "audio") return "音频生成";
    return "图片生成";
}

/** 按节点 ID 查找当前画布节点。 */
export function findNode(state: CanvasSnapshot | null, id: string): CanvasNode | undefined {
    return (state?.nodes || []).find((node) => node.id === id);
}

/** 移除对象中未设置的生成参数。 */
export function cleanRecord(value: Record<string, unknown>) {
    return Object.fromEntries(Object.entries(value).filter(([, item]) => item !== undefined && item !== ""));
}

File: mcp/src/canvas_mcp/operations/shared.py
"""Shared builders for translating high-level tools into ``app_apply_ops``."""

from __future__ import annotations

from typing import Any

from canvas_mcp.types import AgentSnapshot, CanvasNode

CanvasToolRequest = dict[str, Any]


def apply_ops(ops: list[Any]) -> CanvasToolRequest:
    """Wrap a list of canvas ops into the single batch tool request."""
    return {"name": "app_apply_ops", "input": {"ops": ops}}


def omit_none(data: dict[str, Any]) -> dict[str, Any]:
    """Drop ``None``-valued keys, matching how ``JSON.stringify`` drops ``undefined``."""
    return {key: value for key, value in data.items() if value is not None}


def text_node_op(input_data: dict[str, Any], x: float, y: float) -> dict[str, Any]:
    """Build an ``add_node`` op for a text node."""
    return omit_none(
        {
            "type": "add_node",
            "id": input_data.get("id"),
            "nodeType": "text",
            "title": input_data.get("title"),
            "position": {"x": x, "y": y},
            "width": input_data.get("width"),
            "height": input_data.get("height"),
            "metadata": {"content": input_data.get("text") or "", "status": "success", "fontSize": 14},
        }
    )


def config_node_op(node_id: str, input_data: dict[str, Any], x: float, y: float) -> dict[str, Any]:
    """Build an ``add_node`` op for a generation config node."""
    mode = generation_mode(input_data.get("mode"))
    prompt = str(input_data.get("prompt") or "")
    return omit_none(
        {
            "type": "add_node",
            "id": node_id,
            "nodeType": "config",
            "title": str(input_data.get("title") or generation_title(mode)),
            "position": {"x": x, "y": y},
            "width": input_data.get("width"),
            "height": input_data.get("height"),
            "metadata": clean_record(
                {
                    "generationMode": mode,
                    "composerContent": prompt,
                    "prompt": prompt,
                    "status": "idle",
                    "model": input_data.get("model"),
                    "size": input_data.get("size"),
                    "quality": input_data.get("quality"),
                    "count": input_data.get("count"),
                    "seconds": input_data.get("seconds"),
                    "vquality": input_data.get("vquality"),
                    "generateAudio": input_data.get("generateAudio"),
                    "watermark": input_data.get("watermark"),
                    "videoMode": input_data.get("videoMode"),
                    "audioVoice": input_data.get("audioVoice"),
                    "audioFormat": input_data.get("audioFormat"),
                    "audioSpeed": input_data.get("audioSpeed"),
                    "audioInstructions": input_data.get("audioInstructions"),
                }
            ),
        }
    )


def run_generation_op(node_id: str, mode: str, prompt: str | None = None) -> dict[str, Any]:
    """Build a ``run_generation`` op."""
    return omit_none({"type": "run_generation", "nodeId": node_id, "mode": mode, "prompt": prompt})


def generation_mode(value: Any) -> str:
    """Normalise an unknown generation mode to one the canvas supports."""
    return value if value in ("text", "video", "audio") else "image"


def generation_title(mode: str) -> str:
    """Default node title for a generation mode."""
    if mode == "text":
        return "文本生成"
    if mode == "video":
        return "视频生成"
    if mode == "audio":
        return "音频生成"
    return "图片生成"


def find_node(state: AgentSnapshot | None, node_id: str) -> CanvasNode | None:
    """Find a node by id in the current canvas snapshot."""
    return next((node for node in (state or {}).get("nodes") or [] if node.get("id") == node_id), None)


def clean_record(value: dict[str, Any]) -> dict[str, Any]:
    """Remove unset generation parameters (``None`` or empty string)."""
    return {key: item for key, item in value.items() if item is not None and item != ""}
